FoodOrderingApp-api/src/main/java/com/upgrad/FoodOrderingApp/api/controller/RestaurantController.java

```diff
@@ -0,0 +1,336 @@
+package com.upgrad.FoodOrderingApp.api.controller;
+
+
+import com.upgrad.FoodOrderingApp.api.model.*;
+import com.upgrad.FoodOrderingApp.service.businness.*;
+import com.upgrad.FoodOrderingApp.service.entity.CategoryEntity;
+import com.upgrad.FoodOrderingApp.service.entity.CustomerEntity;
+import com.upgrad.FoodOrderingApp.service.entity.ItemEntity;
+import com.upgrad.FoodOrderingApp.service.entity.RestaurantEntity;
+import com.upgrad.FoodOrderingApp.service.exception.AuthorizationFailedException;
+import com.upgrad.FoodOrderingApp.service.exception.CategoryNotFoundException;
+import com.upgrad.FoodOrderingApp.service.exception.InvalidRatingException;
+import com.upgrad.FoodOrderingApp.service.exception.RestaurantNotFoundException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.math.BigDecimal;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.UUID;
+
+// Restaurant Controller Handles all  the restaurant related endpoints
+
+@CrossOrigin
+@RestController
+@RequestMapping("/restaurant")
+public class RestaurantController {
+
+    @Autowired
+    RestaurantService restaurantService; // Handles all the Service Related to Restaurant.
+
+    @Autowired
+    CategoryService categoryService; // Handles all the Service Related to category.
+
+    @Autowired
+    ItemService itemService; // Handles all the Service Related to item.
+
+    @Autowired
+    CustomerService customerService; // Handles all the Service Related to Customer.
+
+
+
+    /* The method handles get All Restaurants request
+    & produces response in RestaurantListResponse and returns list of restaurant with details from the db. If error returns error code and error message.
+    */
+    @CrossOrigin
+    @RequestMapping(method = RequestMethod.GET,path = "",produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<RestaurantListResponse>getAllRestaurants(){
+
+        //Calls restaurantsByRating method of restaurantService to get the list of restaurant entity.
+        List<RestaurantEntity> restaurantEntities = restaurantService.restaurantsByRating();
+
+        //Creating restaurant list for the response
+        List<RestaurantList> restaurantLists = new LinkedList<>();
+        for (RestaurantEntity restaurantEntity : restaurantEntities) { //Looping for each restaurant entity in restaurantEntities
+
+            //Calls  getCategoriesByRestaurant to get categories of the corresponding restaurant.
+            List<CategoryEntity> categoryEntities = categoryService.getCategoriesByRestaurant(restaurantEntity.getUuid());
+            String categories = new String();
```

Comment:
Our static analysis engine detects that your code is problematic. According to the [Performance (PERFORMANCE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#performance-performance), [BC: Impossible downcast of toArray() result (BC_IMPOSSIBLE_DOWNCAST_OF_TOARRAY)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#dm-method-invokes-inefficient-new-string-constructor-dm-string-void-ctor).

Creating a new `java.lang.String` object using the no-argument constructor wastes memory because the object so created will be functionally indistinguishable from the empty string constant `""`.  Java guarantees that identical string constants will be represented by the same `String` object.  Therefore, you should just use the empty string constant directly.